src/main/scala/lms/collection/PointerOps.scala

```diff
@@ -0,0 +1,42 @@
+package lms.collection
+
+import lms.core._
+import lms.util._
+import lms.core.stub._
+import lms.core.Backend._
+import lms.core.virtualize
+import lms.core.utils.time
+import lms.macros.SourceContext
+import scala.collection.mutable.HashMap
+
+trait PointerOps { b: Base =>
+  object Pointer {
+    def apply[A: Manifest](x: Var[A])(implicit pos: SourceContext) = {
+      val a = Wrap[Pointer[A]](Adapter.g.reflect("pointer-new", UnwrapV(x)))
+      cache += ((Unwrap(a), UnwrapV(x)))
+      a
+    }
+    def deref[A: Manifest](x: Rep[Pointer[A]])(implicit pos: SourceContext) = {
```

Comment:
Well, then perhaps it needs a better name. But anyway, with an implicit `PointerOps` class, it would be more natural to write `p.deref` instead `Pointer.deref(p)`.